rp2040-hal/src/gpio/mod.rs

```diff
@@ -233,7 +233,7 @@ impl<I: PinId, F: func::Function, P: PullType> Pin<I, F, P> {
     }
 
     /// Convert the pin from one state to the other.
-    pub fn into<F2, P2>(self) -> Pin<I, F2, P2>
+    pub fn into_typestate<F2, P2>(self) -> Pin<I, F2, P2>
```

Comment:
I stumbled upon this change while porting Neotron code to the new API.
It was not immediately obvious what `into()` needs to be replaced by. Perhaps there should be some migration guide?

At first, I thought that this is a bad change, as `.into()` is a common pattern and easily discoverable. But then, I wondered if it was the correct abstraction in the first place. `.into()` is for type conversion, and usually doesn't have any side effects. Is it an anti-pattern to use `.into()` like this? If yes, then we should also avoid names like `.into_typestate()`, and prefer something like `.set_typestate()`.

If HAL users have to fix all calls to `.into()` anyway, and we already lose the nicely discoverable name, we could as well switch to some "more correct" name. That leaves the `.into_...`-pattern for pure type conversions without side effects.

Reply:
https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv lists `as_`, `into_` and `to_` as common prefixes for type conversion functions. There's no explicit rule that forbids to use those names for something else.

What do other HALs do?

- esp-idf-hal works differently: https://github.com/esp-rs/esp-idf-hal/blob/master/examples/button.rs#L18-L19
- stm32f1xx-hal uses the `into_` prefix: https://github.com/stm32-rs/stm32f1xx-hal/blob/master/examples/blinky.rs#L40
- embassy works more like esp-idf-hal: https://github.com/embassy-rs/embassy/blob/master/examples/rp/src/bin/blinky.rs#LL15C1-L15C53
- nrf-hal uses the `into_`-prefix, https://github.com/nrf-rs/nrf-hal/blob/master/examples/blinky-button-demo/src/main.rs#L22-L23

So using `into_` is widespread, but not universal.

Reply:
According to the API guidelines, the prefix should probably be `to_`:

> Conversions prefixed `to_`, on the other hand, typically stay at the same level of abstraction but do some work to change from one representation to another.

In our case it definitely stays the same level of abstraction. It's the same generic type with different parameters.

Reply:
The three prefixes are meant for functions which do pure type conversions (which might still be expensive, for example when converting strings), but don't perform other actions.

But the GPIO methods also configure the hardware. And that's not just some detail, I'd argue it's the main point of the call. That it also returns a different data type is nice, because it allows to detect some errors at compile time. But it's not the reason the method is called in the first place.

Therefore, IMHO, the functions should have names that don't start with one of these prefixes.

BTW, `.into_typestate()` (and `.set_typestate()`, as I suggested above) don't reflect the intention of the user at all, regardless of the prefix. The actual function is more like `.reconfigure_by_typestate()`  - but that's quite cumbersome.
What about `.reconfigure()`, `.reconfigure_pull_type()`, `.reconfigure_function()`?

Notably, `.into_dyn_pin()` is fine, or could be replaced by `.as_dyn_pin()`, as it really only changes the type, but doesn't change the pin itself.

Of course, we might still decide to keep the `into_` prefix for consistency with several other HALs. Especially as many of the functions, like `.into_pull_up_input()` are quite good at discoverability and readability. But I really worry about `.into_typestate()`. It's both difficult to find if you do not yet know the name of the function, and difficult to understand if you read the code.

Reply:
I wouldn't consider the pin's state separate from the type.
That's why we type state it and all the effort to guaranty the unicity of the pin IDs.

So we are effectively converting (with some reconfiguration cost) a pin from one state to another. This matches the type conversion of `into_` for owned to owned (non-copy) types.

Reply:
Leaving this here as I'm reading them:
https://willcrichton.net/rust-api-type-patterns/state_machines.html
https://docs.rust-embedded.org/book/static-guarantees/state-machines.html

Reply:
> I stumbled upon this change while porting Neotron code to the new API.
It was not immediately obvious what into() needs to be replaced by. Perhaps there should be some migration guide?

We could keep into and mark it `#[deprecated]` to help transition to the new API.

> At first, I thought that this is a bad change, as .into() is a common pattern and easily discoverable. But then, I wondered if it was the correct abstraction in the first place. .into() is for type conversion, and usually doesn't have any side effects. Is it an anti-pattern to use .into() like this? If yes, then we should also avoid names like .into_typestate(), and prefer something like .set_typestate().

Since the type is a representation of the state of the Pin, it don't consider the configuration of the hardware on type transition as a side effect. It's the actual purpose of these type-states.

> If HAL users have to fix all calls to .into() anyway, and we already lose the nicely discoverable name, we could as well switch to some "more correct" name. That leaves the .into_...-pattern for pure type conversions without side effects.

I couldn't find a recurrent pattern for type-state transitions. They usually use verbs corresponding to the actual transition.
I couldn't think of alternative names that would express these better.
- `set_function<F>()`/`set_pull_type<P>()`/`set_state<F, P>()`
- `into_function<F>()`/`into_pull_type<P>()`/`into_state<F, P>()`
- `to_function<F>()`/`to_pull_type<P>()`/`to_state<F, P>()`
- ?

Reply:
I just wanted to jump in and say: I enjoyed reading the discussion above. Congratulations to all for Being Nice On The Internet :)